src/ngraph/ngraph_emitter.cc

```diff
@@ -923,112 +925,99 @@ void Emitter::CreateLayerOps() {
-                            return ngraph::builder::variance(node, axes);
-                          });
-      ngraph::AxisVector order(ng_mean->get_shape().size());
-      std::iota(order.begin(), order.end(), 0);
-      auto ng_mean_temp = std::make_shared<ngraph::op::Reshape>(
-          ng_mean, order, ng_in_moving_mean->get_shape());
-      auto ng_var_temp = std::make_shared<ngraph::op::Reshape>(
-          ng_var, order, ng_in_moving_var->get_shape());
-
-      // update running averages
-
-      NgraphNodePtr ng_one = makeConstant(ng_in_moving_mean->get_element_type(),
-                                          ng_in_moving_mean->get_shape(), "1");
-      NgraphNodePtr ng_momentum =
-          makeConstant(ng_in_moving_var->get_element_type(),
-                       ng_in_moving_var->get_shape(), std::to_string(momentum));
-      ngraph::Shape s = ng_in_moving_mean->get_shape();
+      const NgraphNodePtr ng_momentum =
+        makeConstant(ng_in_moving_var->get_element_type(),
+            ng_in_moving_var->get_shape(), momentum);
 
       aux_op_map_[node->inputs_[kMovingMean]] =
-          ng_in_moving_mean * ng_momentum +
-          ng_mean_temp * (ng_one - ng_momentum);
+          ng_in_moving_mean * ng_momentum + ng_batch_mean * (ng_one - ng_momentum);
+
       aux_op_map_[node->inputs_[kMovingVar]] =
-          ng_in_moving_var * ng_momentum + ng_var_temp * (ng_one - ng_momentum);
+          ng_in_moving_var * ng_momentum + ng_batch_var * (ng_one - ng_momentum);
 
-    } else {
-      // we expect to use global stats with inference
-      ng_mean = std::make_shared<ngraph::op::Reshape>(
-          ng_in_moving_mean, convert_order, convert_shape);
-      ng_var = std::make_shared<ngraph::op::Reshape>(
-          ng_in_moving_var, convert_order, convert_shape);
+      return ng_normalized_data;
     }
 
-    NgraphNodePtr ng_eps = makeConstant(
-        ng_var->get_element_type(), ng_var->get_shape(), std::to_string(eps));
-    NgraphNodePtr denom = std::make_shared<ngraph::op::Sqrt>(ng_var + ng_eps);
-
-    NgraphNodePtr numerator =
-        make_with_numpy_broadcast<ngraph::op::Subtract>(ng_in_data, ng_mean);
-
-    NgraphNodePtr result =
-        make_with_numpy_broadcast<ngraph::op::Divide>(numerator, denom);
-
-    ng_in_gamma = std::make_shared<ngraph::op::Reshape>(
-        ng_in_gamma, convert_order, convert_shape);
-    ng_in_beta = std::make_shared<ngraph::op::Reshape>(
-        ng_in_beta, convert_order, convert_shape);
+    //----------------------------------------------------------------------------------------------
+    // Hybrid mode: use externally supplied mean/variance (as with inference), but also allow
+    // autodifferentiation (as with training).
+    //----------------------------------------------------------------------------------------------
+    if ((exe_mode_ == GraphExeMode::kTrain) && (use_global_stats)) {
+      CHECK(false && "Hybrid-mode batchnorm not implemented yet.");
+      return {};
```

Comment:
I think we should use `create_batchnorm_inference_without_ngraph_bn_op` and autodiff it?

Reply:
I thought @pruthviIntel put in a PR for doing training like this with the op?

Reply:
Yeah, I've already worked out how to do this with and without nGraphs' BN op.  But none of our test cases seems to exercise this, and I wanted to get the PR out ASAP since it's been a while.

Reply:
It's tested in tests/python/gpu/test_operator_gpu.py

Reply:
> It's tested in tests/python/gpu/test_operator_gpu.py

When I ran that the other day, I didn't see my handler for this case get tickled.  Will double check.